src/game/entities/swarm.ts

```diff
@@ -0,0 +1,62 @@
+import { Sprite, SpriteClass } from "kontra";
+import { theme } from "../../resource/theme";
+
+type SwarmProps = {
+    mass: number,
+    stationedAt: CelestialClass|null,
+} & Partial<Parameters<typeof Sprite>[0]>;
+
+export class SwarmClass extends SpriteClass {
+    private mass: string;
+    private stationedAt: CelestialClass|null;
+    
+    constructor(props: Partial<SwarmProps>) {
+        super(props);
+        this.mass = props.mass;
+        this.stationedAt = props.stationedAt;
+    }
+
+    update(dt: number): void {
+        super.update();
+        if (this.stationedAt !== null) {
+            this.devourCelestial(dt);
+        }
+    }
+
+    draw(): void {
+        if (this.stationedAt === null) {
+            this.drawInSpace();
+        } else {
+            this.drawOnSurface();         
+        }
+        this.drawLabels();
+    }
+
+    private drawInSpace() {
+        // TODO: add render logic for space
+    }
+
+    private drawOnSurface() {
+        // TODO: add render logic on the surface
+    }
+
+    private drawLabels() {
+        // TODO: add render logic for text labels with swarm info. For now, we utilize debug UI
+    }
+
+    private devourCelestial(dt: number) {
+        const deltaMass = Math.min(
+            this.stationedAt.mass,
+            this.mass * this.stationedAt.massAvailability * dt
+        );
+
+        this.mass += deltaMass;
+        this.stationedAt.mass -= deltaMass;
```

Comment:
Maybe it is better to move this logic to higher level of abstraction? 
```js
universe.iterate = () => {
  const calcMass = (celestial, swarms[]) => {
    const swarmAtCelestial = getSwarm(celestial.coordinates);
    return swarmAtCelestial == null ? celectial.mass : celestial.mass - swarm.eatingSpeed
  }
  const newCelestials = oldCelestials.forEach(c => c,mass = calcMass(c, swarms))
}
```

Reply:
I would argue that the logic of "swarm consuming the mass and converting it to more swarm" belongs in the swarm class. One could extract it into some other aggregating type, but the next step is to introduce some way of moving the swarms around. Would that also live outside the swarm class? I'd prefer to leave it where it is.

Reply:
Ok, let’s swarm be the main class that will rule other classes